### PR TITLE
crimson/osd: ShardServices::dispatch_context_transaction to flush

### PR DESCRIPTION
**Pull request by cyx1231st**
Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

A majority of empty transactions come from `ShardServices::dispatch_context_transaction` when the OSD is idle.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/shard_services.cc [hunk 1]
```diff
@@ -106,12 +106,17 @@ seastar::future<> ShardServices::send_to_osd(
...
   crimson::os::CollectionRef col, PeeringCtx &ctx) {
-  logger().debug("ShardServices::dispatch_context_transaction: do_transaction...");
-  auto ret = store.do_transaction(
-    col,
-    std::move(ctx.transaction));
-  ctx.reset_transaction();
-  return ret;
+  if (ctx.transaction.empty()) {
+    logger().debug("ShardServices::dispatch_context_transaction: flush...");
+    return store.flush(col);
```
Comment (cyx1231st) [inline src/crimson/osd/shard_services.cc:111]: Do we need to flush here if the transaction is empty?

Comment (athanatos) [reply src/crimson/osd/shard_services.cc:111]: No, if the transaction is actually empty here I don't think we need to do anything.

**Review by athanatos — APPROVED** [review]

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
```
The following tests FAILED:
	 53 - unittest_bufferlist (Failed)
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
```
The following tests FAILED:
	  4 - run-tox-mgr (Failed)
```

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Pull request merged**